src/Nethermind/Nethermind.Xdc/SnapshotManager.cs

```diff
@@ -1,87 +1,135 @@
+        }
 
-    public bool IsValidVote(Snapshot snapshot, Address address, bool authorize)
-    {
-        throw new NotImplementedException();
+        Span<byte> key = hash.Bytes;
+        if (!_snapshotDb.KeyExists(key))
+            return null;
+        Span<byte> value = _snapshotDb.Get(key);
+        if (value.IsEmpty)
+            return null;
+        SnapshotDecoder snapshotDecoder = new();
+
+        var decoded = snapshotDecoder.Decode(value, RlpBehaviors.None);
+        snapshot = decoded;
+        _snapshotsByHash.Set(hash, snapshot);
+        return snapshot;
     }
 
-    public void TryCacheSnapshot(Snapshot snapshot)
+    public Snapshot? GetSnapshotByHeader(XdcBlockHeader? header)
     {
-        throw new NotImplementedException();
+        if (header is null)
+            return null;
+        return GetSnapshot(header.Hash);
     }
 
-    public bool TryGetSnapshot(long number, Hash256 hash, out Snapshot snapshot)
+    public Snapshot? GetSnapshotByHeaderNumber(ulong number, ulong xdcEpoch, ulong xdcGap)
     {
-        throw new NotImplementedException();
+        ulong gapBlockNum = Math.Max(0, number - number % xdcEpoch - xdcGap);
+
+        return GetSnapshotByGapNumber(gapBlockNum);
     }
 
-    public bool TryGetSnapshot(XdcBlockHeader header, out Snapshot snapshot)
+
+    public Snapshot? GetSnapshotByGapNumber(ulong gapBlockNum)
     {
-        throw new NotImplementedException();
+        Hash256 gapBlockHash = _tree.FindHeader((long)gapBlockNum)?.Hash;
+
+        if (gapBlockHash is null)
+            return null;
+
+        return GetSnapshot(gapBlockHash);
     }
 
-    public bool TryGetSnapshot(ulong gapNumber, bool isGapNumber, out Snapshot snap)
+    public bool StoreSnapshot(Snapshot snapshot)
     {
-        throw new NotImplementedException();
+        if (snapshot is null)
+            return false;
+        Span<byte> key = snapshot.HeaderHash.Bytes;
+
+        if (_snapshotDb.KeyExists(key))
+            return false;
+
+        var rlpEncodedSnapshot = _snapshotDecoder.Encode(snapshot, RlpBehaviors.None);
+
+        _snapshotDb.Set(key, rlpEncodedSnapshot.Bytes);
+        return true;
     }
 
-    public bool TryStoreSnapshot(Snapshot snapshot)
+    public Address[] CalculateNextEpochMasternodes(XdcBlockHeader xdcHeader)
     {
-        throw new NotImplementedException();
+        Snapshot snapshot = GetSnapshotByHeader(xdcHeader);
+        if (snapshot is null)
+            throw new InvalidOperationException($"No snapshot found for header {xdcHeader.Number}:{xdcHeader.Hash.ToShortString()}");
+
+        Address[] masternodes = new Address[snapshot.MasterNodes.Length - snapshot.PenalizedNodes.Length];
+
+        int index = 0;
+        foreach (var addr in snapshot.MasterNodes)
+        {
+            if (snapshot.PenalizedNodes.Contains(addr))
+            {
+                continue;
+            }
+
+            masternodes[index++] = addr;
+        }
+        return masternodes;
```

Comment:
This is a `.Where()` right?

```suggestion
        Address[] masternodes = snapshot.MasterNodes
            .Where(addr => !snapshot.PenalizedNodes.Contains(addr))
            .ToArray();

          return masternodes;
```

Reply:
I avoided Linq, because i am not sure of the perf penalty it will have